### PR TITLE
Fix WebSockets receiving messages in chunks does not detect correct message size

### PR DESCRIPTION
**Pull request by caesar1995**
Fix: #28097

It's a bug in `WebScoketBuffer`, we forget to decrease the _bufferedPayloadReceiveResult.Count.
Framework [doesn't has this problem.](https://github.com/Microsoft/referencesource/blob/4fe4349175f4c5091d972a7e56ea12012f1e7170/System/net/System/Net/WebSockets/WebSocketReceiveResult.cs#L45)

Verified with the repro provided by customer.

### CODE REVIEW

File: src/System.Net.WebSockets/ref/System.Net.WebSockets.cs [hunk 1]
```diff
@@ -117,7 +117,7 @@ public partial class WebSocketReceiveResult
         public WebSocketReceiveResult(int count, System.Net.WebSockets.WebSocketMessageType messageType, bool endOfMessage, System.Nullable<System.Net.WebSockets.WebSocketCloseStatus> closeStatus, string closeStatusDescription) { }
         public System.Nullable<System.Net.WebSockets.WebSocketCloseStatus> CloseStatus { get { throw null; } }
         public string CloseStatusDescription { get { throw null; } }
-        public int Count { get { throw null; } }
+        public int Count { get { throw null; } set { } }
```
Comment (stephentoub) [inline src/System.Net.WebSockets/ref/System.Net.WebSockets.cs]: We should not be adding APIs as part of this.

File: src/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketReceiveResult.cs [hunk 2]
```diff
@@ -31,7 +31,7 @@ public WebSocketReceiveResult(int count, WebSocketMessageType messageType, bool
             CloseStatusDescription = closeStatusDescription;
         }
 
-        public int Count { get; }
+        public int Count { get; set;}
```
Comment (rmkerr) [inline src/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketReceiveResult.cs]: Is this changing a public API? It also doesn't seem like this is something that should be publicly set, based off of the documentation.

File: src/System.Net.WebSockets/ref/System.Net.WebSockets.cs (hunk 1 above)
Comment (caesar1995) [reply src/System.Net.WebSockets/ref/System.Net.WebSockets.cs]: I will change to follow the Framework's pattern.

**Comment by caesar1995** [comment]
The `WebSocketBuffer` class is in `System.Net.HttpListener` project, and the `WebSocketReceiveResult` class is in `System.Net.WebSocket` project.

We want to keep track/change the value of `Count` field in `WebSocketReceiveResult` from `WebSocketBuffer` class, without introducing new API.

Here are the approaches I tried:

1. Follow Framework's pattern, create an internal (visible to same assembly) `Copy()` method in `WebSocketReceiveResult`. But we are not able to detecet the method from `WebSocketBuffer` class. (In Framework, these two classes are from same assembly)
2. I need to take the second approach, which is creating new `WebSocketReceiveResult` with new `Count` value, to replace the old one.

Please let me know if there is better way to achieve this. (I have considered using reflection, but I don't think it's a good practice to do so)

**Comment by stephentoub** [comment]
> Please let me know if there is better way to achieve this.

How about just making `_bufferedPayloadReceiveResult` something other than a `WebSocketReceiveResult`?  It's private and it doesn't look like the instance is handed out to anyone; seems like it was just written this way for convenience, but you could either create your own type, or just expand to have the fields needed on WebSocketBuffer directly.

**Comment by caesar1995** [comment]
> but you could either create your own type, or just expand to have the fields needed on WebSocketBuffer directly.

I choose to create own type to group up related fields. And I used nested class instead of struct because the type is not immutable, we need to change the `Count` filed frequently, and the type is not logically representing a single value.

File: src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/WebSocketBuffer.cs [hunk 3]
```diff
@@ -326,12 +326,12 @@ internal bool ReceiveFromBufferedPayload(ArraySegment<byte> buffer, out WebSocke
 
             int bytesTransferred = Math.Min(buffer.Count, _bufferedPayloadReceiveResult.Count);
 
+            _bufferedPayloadReceiveResult.Count = _bufferedPayloadReceiveResult.Count - bytesTransferred;
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/WebSocketBuffer.cs]: Nit: could use -=

Comment (caesar1995) [reply src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/WebSocketBuffer.cs]: Oh, will fix.

**Pull request merged**